src/python/pants/core/util_rules/posix_tools.py

```diff
@@ -0,0 +1,59 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import ClassVar, TypeVar
+
+from pants.engine.process import BinaryPath, BinaryPathRequest, BinaryPaths, BinaryPathTest
+from pants.engine.rules import Get, collect_rules, rule
+
+T = TypeVar("T", bound="PosixBinary")
+
+
+class PosixBinary(BinaryPath):
+
+    binary_name: ClassVar[str]
+    binary_test: ClassVar[BinaryPathTest | None]
+
+    @classmethod
+    def using(cls: type[T], binary_path: BinaryPath) -> T:
+        return cls(path=binary_path.path, fingerprint=binary_path.fingerprint)
+
+
+@dataclass(frozen=True)
+class PosixBinaryRequest:
+    type: type[PosixBinary]
+
+
+class LnBinary(PosixBinary):
+    binary_name = "ln"
+    binary_test = None
+
+
+SEARCH_PATHS = ("/usr/bin", "/bin", "/usr/local/bin")
+
+
+@rule
+async def find_ln() -> LnBinary:
+    ln = await Get(BinaryPath, PosixBinaryRequest(LnBinary))
+    return LnBinary.using(ln)
+
+
+@rule
+async def find_posix_binary(request: PosixBinaryRequest) -> BinaryPath:
+    path_request = BinaryPathRequest(
+        binary_name=request.type.binary_name,
+        search_path=SEARCH_PATHS,
+        test=request.type.binary_test,
+    )
+    paths = await Get(BinaryPaths, BinaryPathRequest, path_request)
+    first_path = paths.first_path_or_raise(
+        path_request, rationale=f"use `{path_request.binary_name}` in internal shell scripts"
+    )
+    return BinaryPath(first_path.path, first_path.fingerprint)
+
+
```

Comment:
This is probably worth generalizing, but it would be good to lay some ground rules on which tools should be declared in this file, vs elsewhere. For example, `src/python/pants/core/util_rules/archive.py` already contains archive-specific tools (and probably should), including some that are part of `posix`. And `BinaryPathRequest` is declared directly in `process.py`.

As declared, these rules can't be used with `BashBinaryRequest`, because that would cause a cycle with `process.py`. So maybe they could both be moved into `pants.engine.shell` or `pants.engine.scripting` or something? Meanwhile, another odd duck is `pants.python.binaries`, which defines `PythonBinary`.

Reply:
@stuhood Is `BashBinaryRequest` ever used instead of `BashBinary`?

Reply:
Imo we should move `unzip` and friends to this file. This is more appropriate.

Reply:
I don't think so... probably because it never requires laziness the way that some of the other tools do (via `await Get(Tool, ToolRequest())` rather than putting `Tool` directly in a `@rule` argument list).

Reply:
OK. So the `BashBinaryRequest` cycle difficulty is theoretical then, yeah?

At the moment, I think the rationale I had for making this file is that it was for miscellaneous tools that don't fit elsewhere, where the difficulty of making a `*Binary` object (and making the shell script safer in the process) was made unnecessarily high by virtue of there not being a file where it fit.

A separate tool for archive binaries makes sense. I don't have an issue with the Bash compat being in yet another file. The benefit here is that if there isn't a file where a given binary path object should go, this is that file. Maybe the name is too specific? Maybe it could be `miscellaneous_binaries.py` or something like that?